### ISSUE TITLE
3.8.0 build failure

### ISSUE THREAD

**Issue by abelbeck** [pciutils/pciutils#98]
Using x86_64-unknown-linux-gnu-gcc (crosstool-NG 1.24.0) 6.5.0

`filter.c` fails to compile:
```
make[2]: Entering directory '/home/dev/astlinux/trunk/output/build/pciutils-3.8.0/lib'
/home/dev/astlinux/trunk/output/host/usr/bin/x86_64-unknown-linux-gnu-gcc -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o init.o init.c
/home/dev/astlinux/trunk/output/host/usr/bin/x86_64-unknown-linux-gnu-gcc -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o access.o access.c
/home/dev/astlinux/trunk/output/host/usr/bin/x86_64-unknown-linux-gnu-gcc -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o generic.o generic.c
/home/dev/astlinux/trunk/output/host/usr/bin/x86_64-unknown-linux-gnu-gcc -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o dump.o dump.c
/home/dev/astlinux/trunk/output/host/usr/bin/x86_64-unknown-linux-gnu-gcc -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o names.o names.c
/home/dev/astlinux/trunk/output/host/usr/bin/x86_64-unknown-linux-gnu-gcc -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -pipe -Os  -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o filter.o filter.c
{standard input}: Assembler messages:
{standard input}:5: Error: multiple versions [`pci_filter_init@@LIBPCI_3.8'|`pci_filter_init@LIBPCI_3.3'] for symbol `pci_filter_init_v38'
{standard input}:7: Error: multiple versions [`pci_filter_parse_slot@@LIBPCI_3.8'|`pci_filter_parse_slot@LIBPCI_3.3'] for symbol `pci_filter_parse_slot_v38'
{standard input}:9: Error: multiple versions [`pci_filter_parse_id@@LIBPCI_3.8'|`pci_filter_parse_id@LIBPCI_3.3'] for symbol `pci_filter_parse_id_v38'
{standard input}:11: Error: multiple versions [`pci_filter_match@@LIBPCI_3.8'|`pci_filter_match@LIBPCI_3.3'] for symbol `pci_filter_match_v38'
make[2]: *** [<builtin>: filter.o] Error 1
```
I tried adding `-std=c99` but that caused other errors.

Any help would be appreciated.

**Comment by abelbeck** [comment]
As a temporary fix, at the end of `filter.c`, I removed
```
SYMBOL_VERSION(pci_filter_init_v38, pci_filter_init@LIBPCI_3.3);
SYMBOL_VERSION(pci_filter_parse_slot_v38, pci_filter_parse_slot@LIBPCI_3.3);
SYMBOL_VERSION(pci_filter_parse_id_v38, pci_filter_parse_id@LIBPCI_3.3);
SYMBOL_VERSION(pci_filter_match_v38, pci_filter_match@LIBPCI_3.3);
```
since all references in our build are `LIBPCI_3.8` ... it builds and works.

Possibly the real fix is to create stubs for `pci_filter_init_v33` calling `pci_filter_init_v38`, etc. and then
```
SYMBOL_VERSION(pci_filter_init_v33, pci_filter_init@LIBPCI_3.3);
SYMBOL_VERSION(pci_filter_parse_slot_v33, pci_filter_parse_slot@LIBPCI_3.3);
SYMBOL_VERSION(pci_filter_parse_id_v33, pci_filter_parse_id@LIBPCI_3.3);
SYMBOL_VERSION(pci_filter_match_v33, pci_filter_match@LIBPCI_3.3);
```

**Comment by mikaku** [comment]
Hi,

I've also experienced a new problem when building 3.8.0 on my [FiwixOS](https://www.fiwix.org) (3.8.7 version was built successfully):

```
ls-ecaps.c: In function 'cap_rcec':
ls-ecaps.c:657:7: error: 'for' loop initial declarations are only allowed in C99 mode
ls-ecaps.c:657:7: note: use option -std=c99 or -std=gnu99 to compile your code
```

https://github.com/pciutils/pciutils/blob/d224993d423bcd0e88d8eacc0464bc3a66972e95/ls-ecaps.c#L657

Adding the flag `-std=c99` to `gcc` is not an option as it causes other errors, so the only way in my case was to declare `dev` out from the `for` loop.

Please, consider keep compatibility with C89.

**Comment by pali** [comment]
@mikaku: FYI following two external commit fixes that compatibility issues: https://github.com/pali/pciutils/commit/c63204d86d1ee8531daf3d1cfe02df33c2756bac https://github.com/pali/pciutils/commit/a4bb46627f67b6af7a257c74968379aafe56b087

**Comment by gollux** [comment]
@abelbeck Which commit are you compiling? I hoped I fixed these just before releasing 3.8.0.

@mikaku What "other errors" arise with `-std=c99`?

**Comment by abelbeck** [comment]
@gollux : I was compiling the 3.8.0 release.

**Comment by mikaku** [comment]
> @mikaku: FYI following two external commit fixes that compatibility issues: [pali@c63204d](https://github.com/pali/pciutils/commit/c63204d86d1ee8531daf3d1cfe02df33c2756bac) [pali@a4bb466](https://github.com/pali/pciutils/commit/a4bb46627f67b6af7a257c74968379aafe56b087)

Yes, I did a similar patch to fix the compatibility issue.
Thanks.

**Comment by mikaku** [comment]
> @mikaku What "other errors" arise with `-std=c99`?

@gollux, the following is the built transcription after including the `-std=c99` parameter:

```
cd lib && ./configure
Configuring libpci for your system... i386--fiwix 1.3.0 i386 fiwix
Looking for access methods... i386-ports dump
Checking for zlib support... yes (auto-detected)
Checking for DNS support... no (auto-detected)
Checking whether to build a shared library... no (set manually)
make -C lib all
make[1]: Entering directory '/mnt/builds/pciutils-3.8.0/lib'
cc -O2 -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -std=c99  -c -o init.o init.c
In file included from internal.h:34:0,
                 from init.c:14:
sysdep.h:88:2: error: #error "BYTE_ORDER not defined for your platform"
<builtin>: recipe for target 'init.o' failed
make[1]: *** [init.o] Error 1
make[1]: Leaving directory '/mnt/builds/pciutils-3.8.0/lib'
Makefile:70: recipe for target 'lib/libpci.a' failed
make: *** [lib/libpci.a] Error 2

 *** An error occurred. Build has been stopped ***
```

This error does not appear if the parameter `-std=c99` is not included.

**Comment by michaelolbrich** [comment]
The original problem has nothing to do with the C standard. The other errors occur when `-std=cXX` is used instead of the default `-std=gnuXX`.

From my tests, the `multiple versions ...` errors occur with gcc 8 and 9 (and probably older versions). Building with gcc 10 or 11 works.

**Comment by idc77** [comment]
I can not confirm that building with gcc-11 works
building on gentoo
```
gcc-config -l
 [1] x86_64-pc-linux-gnu-7.3.0
 [2] x86_64-pc-linux-gnu-8.2.0
 [3] x86_64-pc-linux-gnu-8.3.0
 [4] x86_64-pc-linux-gnu-9.1.0
 [5] x86_64-pc-linux-gnu-9.2.0
 [6] x86_64-pc-linux-gnu-9.3.0
 [7] x86_64-pc-linux-gnu-10.2.0
 [8] x86_64-pc-linux-gnu-11.3.0 *
```
```
 * Package:    sys-apps/pciutils-3.8.0
 * Repository: gentoo
 * Maintainer: base-system@gentoo.org
 * USE:        abi_x86_64 amd64 elibc_glibc kernel_linux kmod udev userland_GNU zlib
 * FEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
>>> Unpacking source...
>>> Unpacking pciutils-3.8.0.tar.gz to /var/tmp/portage/sys-apps/pciutils-3.8.0/work
>>> Source unpacked in /var/tmp/portage/sys-apps/pciutils-3.8.0/work
>>> Preparing source in /var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0 ...
 * Will copy sources from /var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0
 * abi_x86_64.amd64: copying to /var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0-abi_x86_64.amd64
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0 ...
 * abi_x86_64.amd64: running multilib-minimal_abi_src_configure
>>> Source configured.
>>> Compiling source in /var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0 ...
 * abi_x86_64.amd64: running multilib-minimal_abi_src_compile
make -j8 -l8 HOST=x86_64-pc-linux-gnu CROSS_COMPILE=x86_64-pc-linux-gnu- CC=x86_64-pc-linux-gnu-gcc AR=x86_64-pc-linux-gnu-ar PKG_CONFIG=x86_64-pc-linux-gnu-pkg-config RANLIB=x86_64-pc-linux-gnu-ranlib DNS=no IDSDIR=$(SHAREDIR)/hwdata MANDIR=$(SHAREDIR)/man PREFIX=/usr SHARED=yes STRIP= ZLIB=yes PCI_COMPRESSED_IDS=0 PCI_IDS=pci.ids LIBDIR=${PREFIX}/lib64 LIBKMOD=yes HWDB=yes 'OPT=-O2 -pipe -march=native' all 
cd lib && ./configure
Configuring libpci for your system... x86_64--linux  x86_64 linux
Looking for access methods... sysfs proc i386-ports dump
Checking for zlib support... yes (set manually)
Checking for DNS support... no (set manually)
Checking for libkmod... yes (set manually)
Checking for udev hwdb support... yes (set manually)
Checking whether to build a shared library... yes (set manually)
make -C lib all
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o lspci.o lspci.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-vpd.o ls-vpd.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-caps.o ls-caps.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-caps-vendor.o ls-caps-vendor.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-ecaps.o ls-ecaps.c
make[1]: Entering directory '/var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0-abi_x86_64.amd64/lib'
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o init.o init.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes   -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-kernel.o ls-kernel.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o access.o access.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-tree.o ls-tree.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o ls-map.o ls-map.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o generic.o generic.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o common.o common.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o dump.o dump.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o setpci.o setpci.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o example.o example.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o names.o names.c
M=`echo 2022-04-18 | sed 's/-01-/-January-/;s/-02-/-February-/;s/-03-/-March-/;s/-04-/-April-/;s/-05-/-May-/;s/-06-/-June-/;s/-07-/-July-/;s/-08-/-August-/;s/-09-/-September-/;s/-10-/-October-/;s/-11-/-November-/;s/-12-/-December-/;s/\(.*\)-\(.*\)-\(.*\)/\3 \2 \1/'` ; sed <lspci.man >lspci.8 "s/@TODAY@/$M/;s/@VERSION@/pciutils-3.8.0/;s#@IDSDIR@#/usr/share/hwdata#;s#@PCI_IDS@#pci.ids#"
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o filter.o filter.c
M=`echo 2022-04-18 | sed 's/-01-/-January-/;s/-02-/-February-/;s/-03-/-March-/;s/-04-/-April-/;s/-05-/-May-/;s/-06-/-June-/;s/-07-/-July-/;s/-08-/-August-/;s/-09-/-September-/;s/-10-/-October-/;s/-11-/-November-/;s/-12-/-December-/;s/\(.*\
)-\(.*\)-\(.*\)/\3 \2 \1/'` ; sed <setpci.man >setpci.8 "s/@TODAY@/$M/;s/@VERSION@/pciutils-3.8.0/;s#@IDSDIR@#/usr/share/hwdata#;s#@PCI_IDS@#pci.ids#"
M=`echo 2022-04-18 | sed 's/-01-/-January-/;s/-02-/-February-/;s/-03-/-March-/;s/-04-/-April-/;s/-05-/-May-/;s/-06-/-June-/;s/-07-/-July-/;s/-08-/-August-/;s/-09-/-September-/;s/-10-/-October-/;s/-11-/-November-/;s/-12-/-December-/;s/\(.*\
)-\(.*\)-\(.*\)/\3 \2 \1/'` ; sed <pcilib.man >pcilib.7 "s/@TODAY@/$M/;s/@VERSION@/pciutils-3.8.0/;s#@IDSDIR@#/usr/share/hwdata#;s#@PCI_IDS@#pci.ids#"
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o names-hash.o names-hash.c
M=`echo 2022-04-18 | sed 's/-01-/-January-/;s/-02-/-February-/;s/-03-/-March-/;s/-04-/-April-/;s/-05-/-May-/;s/-06-/-June-/;s/-07-/-July-/;s/-08-/-August-/;s/-09-/-September-/;s/-10-/-October-/;s/-11-/-November-/;s/-12-/-December-/;s/\(.*\
)-\(.*\)-\(.*\)/\3 \2 \1/'` ; sed <pci.ids.man >pci.ids.5 "s/@TODAY@/$M/;s/@VERSION@/pciutils-3.8.0/;s#@IDSDIR@#/usr/share/hwdata#;s#@PCI_IDS@#pci.ids#"
sed <update-pciids.sh >update-pciids "s@^DEST=.*@DEST=/usr/share/hwdata/pci.ids@;s@^PCI_COMPRESSED_IDS=.*@PCI_COMPRESSED_IDS=0@"
chmod +x update-pciids
M=`echo 2022-04-18 | sed 's/-01-/-January-/;s/-02-/-February-/;s/-03-/-March-/;s/-04-/-April-/;s/-05-/-May-/;s/-06-/-June-/;s/-07-/-July-/;s/-08-/-August-/;s/-09-/-September-/;s/-10-/-October-/;s/-11-/-November-/;s/-12-/-December-/;s/\(.*\
)-\(.*\)-\(.*\)/\3 \2 \1/'` ; sed <update-pciids.man >update-pciids.8 "s/@TODAY@/$M/;s/@VERSION@/pciutils-3.8.0/;s#@IDSDIR@#/usr/share/hwdata#;s#@PCI_IDS@#pci.ids#"
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o names-parse.o names-parse
.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o names-net.o names-net.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden  -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE  -c -o names-cache.o names-cache
.c
{standard input}: Assembler messages:
{standard input}:6: Error: multiple versions [`pci_filter_init@@LIBPCI_3.8'|`pci_filter_init@LIBPCI_3.3'] for symbol `pci_filter_init_v38'
{standard input}:8: Error: multiple versions [`pci_filter_parse_slot@@LIBPCI_3.8'|`pci_filter_parse_slot@LIBPCI_3.3'] for symbol `pci_filter_parse_slot_v38'
{standard input}:10: Error: multiple versions [`pci_filter_parse_id@@LIBPCI_3.8'|`pci_filter_parse_id@LIBPCI_3.3'] for symbol `pci_filter_parse_id_v38'
{standard input}:12: Error: multiple versions [`pci_filter_match@@LIBPCI_3.8'|`pci_filter_match@LIBPCI_3.3'] for symbol `pci_filter_match_v38'
make[1]: *** [<builtin>: filter.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/var/tmp/portage/sys-apps/pciutils-3.8.0/work/pciutils-3.8.0-abi_x86_64.amd64/lib'
make: *** [Makefile:70: lib/libpci.so.3.8.0] Error 2
make: *** Waiting for unfinished jobs....
```

**Comment by tunerooster** [comment]
Same problem here...

What is the solution?

**Comment by w1ll14m** [comment]
Discusstion on kernel.org suggest this might be a binutils bug, I was able to compile pciutils-3.8.0 after I changed from binutils 2.32 to 2.35.

Source:
https://lore.kernel.org/linux-pci/87mtfm7v58.fsf@tarshish/

**Comment by aliceinwire** [comment]
I had the same errors, and I can confirm that switching to binutils from 2.33.1 to 2.38 could make it build correctly.

**Comment by spanevin** [comment]
I have the same issue on gentoo
stable binutils version is 2.37 here now

**Comment by spanevin** [comment]
> > I have the same issue on gentoo stable (installed) binutils version is 2.37 here now, active gcc is 9.3.0 switching to gcc 10 or gcc 11 doesn't help, same error
> 
> Make sure you run `. /etc/profile` after changing binutils via `binutils-config`.

I didn't know about binutils-config at all... it shows that 2.34 is active

switching to 2.37 + gcc11 works, pciutils compiled successfully

**Comment by aliceinwire** [comment]
I think this issue can be closed

**Comment by spanevin** [comment]
> 

Really? Do you think all users know they should switch to newer binutils/gcc for this?
Maybe I don't remember it correctly, but I don't think I've seen any news about this breaking change after portage sync update.

**Comment by spanevin** [comment]
do you understand that depclean don't remove old binutils until you do binutils-config and switch to a new version?

Please read my message above. I use gentoo for more than ten years, but I didn't know/remember that there is binutils-update at all. I only used gcc-update before.

**Comment by thesamesam** [comment]
> do you understand that depclean don't remove old binutils until you do binutils-config and switch to a new version?
> 
> Please read my message above. I use gentoo for more than ten years, but I didn't know/remember that there is binutils-update at all. I only used gcc-update before.

As one of the Portage developers and toolchai maintainers, I can promise you that depcleaning will result in binutils-config automatically choosing an available version instead, otherwise it'd result in dead symlinks. You shouldn't even have such old versions installed, which depcleaning globally (emerge -ac) will handle.

**Comment by spanevin** [comment]
You're right, I thought that a package would be treated as "protected" for depclean if it was an active version... My mistake.

**Comment by idc77** [comment]
Sorry for the offtopic, I never use depclean because I have no confidence that portage would get it right and not delete essential packages.
And yup,
```
 virtual/jre
    selected: 11-r2 
   protected: none 
     omitted: 17 

 virtual/jdk
    selected: 11-r2 
   protected: none 
     omitted: 1.8.0-r6 17 
```
for one. But my keycloak instance runs with jdk11 and it's not negotiable.
So I never bother depcleaning. There's a huge list of packages and it's just too much.

or this
```
 app-editors/nano
    selected: 4.7 
   protected: none 
     omitted: none 
```

anyhow it's really off-topic for this issue so sorry again.
I'll not respond to eventual responses.

On topic: confirm binutils 2.38 and gcc 11 successfully compiles pciutils. Good job finding that needle.

**Comment by thesamesam** [comment]
> Sorry for the offtopic, I never use depclean because I have no confidence that portage would get it right and not delete essential packages. And yup,
> 
> ```
>  virtual/jre
>     selected: 11-r2 
>    protected: none 
>      omitted: 17 
> 
>  virtual/jdk
>     selected: 11-r2 
>    protected: none 
>      omitted: 1.8.0-r6 17 
> ```
> 
> for one. But my keycloak instance runs with jdk11 and it's not negotiable. So I never bother depcleaning. There's a huge list of packages and it's just too much.
> 
> or this
> 
> ```
>  app-editors/nano
>     selected: 4.7 
>    protected: none 
>      omitted: none 
> ```
> 
> anyhow it's really off-topic for this issue so sorry again. I'll not respond to eventual responses.
> 
> On topic: confirm binutils 2.38 and gcc 11 successfully compiles pciutils. Good job finding that needle.

You should then simply add packages you want to world. It's easy.

I suspect your keycloak doesn't correctly hard depend on JDK 11.

As for nano, emerge -vn app-editors/nano. It's because virtual/editor can be satisfied by a bunch of editors, and if another one is installed...

**Comment by lucabonnix** [comment]
The problem is still present with binutils-2.38, but only on armv5 platform (all other architectures with exactly the same libraries/compilers [Slackware-current], including armv7, compile without problems):

Configuring libpci for your system... armv5tel--linux 5.15.15 armv5tel linux [...]
cc -O2 -Wall -W -Wno-parentheses -Wstrict-prototypes -Wmissing-prototypes -fPIC -fvisibility=hidden   -c -o filter.o filter.c
/root/tmp/cc6eQYoj.s: Assembler messages:
/root/tmp/cc6eQYoj.s:16: Error: multiple versions [`pci_filter_init@@LIBPCI_3.8'|`pci_filter_init@LIBPCI_3.3'] for symbol `pci_filter_init_v38'
/root/tmp/cc6eQYoj.s:18: Error: multiple versions [`pci_filter_parse_slot@@LIBPCI_3.8'|`pci_filter_parse_slot@LIBPCI_3.3'] for symbol `pci_filter_parse_slot_v38'
/root/tmp/cc6eQYoj.s:20: Error: multiple versions [`pci_filter_parse_id@@LIBPCI_3.8'|`pci_filter_parse_id@LIBPCI_3.3'] for symbol `pci_filter_parse_id_v38'
/root/tmp/cc6eQYoj.s:22: Error: multiple versions [`pci_filter_match@@LIBPCI_3.8'|`pci_filter_match@LIBPCI_3.3'] for symbol `pci_filter_match_v38'

**Comment by gollux** [comment]
Please try current master, I hope I fixed this in commit 0478e1f3928bfaa34eb910ba2cbaf1dda8f84aab.

**Comment by BKPepe** [comment]
I experienced such behavior while building pciutils for OpenWrt 21.02 and I can confirm that the referenced commit fixes it.
